src/rgw/rgw_process.cc

```diff
@@ -225,7 +225,9 @@ int rgw_process_authenticated(RGWHandler_REST * const handler,
     std::swap(span, s->trace);
   }
   if (ret < 0) {
-    if (s->system_request) {
+    if (!op->check_initialised()) {
+      return ret;
+    } else if (s->system_request) {
```

Comment:
Thanks for reviewing.

Yeah, that's exactly the problem. 
But I'm afraid checking for `-EPERM` and `-EACCES` is not enough. 
If those 2 errors are returned before the operation is fully initialised and we allow to keep going it will crash anyway.

A couple of examples from `RGWCopyObj::verify_permission`: (last line is the line initialising `dest_bucket`)

```c++
  if (identity_policy_res == Effect::Pass && e == Effect::Pass &&
		   !src_acl.verify_permission(this, *s->auth.identity, s->perm_mask,
					      RGW_PERM_READ)) { 
	  return -EACCES;
	}
      //remove src object tags as it may interfere with policy evaluation of destination obj
      if (has_s3_existing_tag || has_s3_resource_tag)
        rgw_iam_remove_objtags(this, s, s->src_object.get(), has_s3_existing_tag, has_s3_resource_tag);

      } else if (!src_acl.verify_permission(this, *s->auth.identity,
					       s->perm_mask,
					    RGW_PERM_READ)) {
	return -EACCES;
      }
    }
  }

  RGWAccessControlPolicy dest_bucket_policy(s->cct);

  if (src_bucket_name.compare(dest_bucket_name) == 0) { /* will only happen if s->local_source
                                                           or intra region sync */
    dest_bucket = src_bucket->clone();
  } else {
    op_ret = store->get_bucket(this, s->user.get(), dest_tenant_name, dest_bucket_name, &dest_bucket, y);
```

Those 2 `-EACCES` would be bypassed, but the `execute` method will crash. 

Adding a new method to the `RGWOp` interface it's maybe not ideal, but the thing is we'd need a way to ensure that bypassing those errors is safe... even if the error is completely related to permissions.

Reply:
Another possible solution would be to re-run the code that initialises the members needed in the `execute` method (not relying on the `init_permissions` method), but that implies double initialisation in many cases and could (maybe) impact performance.